expertiza/expertiza
# E1780. OSS project Teal: Email notification enhancements

E1780. OSS PROJECT TEAL: EMAIL NOTIFICATION ENHANCEMENTS

Credentials for instructor :-
User Id : instructor6
Password : password

Note: When you login as instructor for the first time, change the email id to an accessible one from the Profile tab. When you create any user, please ensure that you give an email id that is accessible to you.

**********************************************************************************

Requirement 1 : 
Original requirement  : Issue #717: When students' accounts are created by importing a CSV file on the Users page, they receive emails with their user-ID and password. But if an account is created by adding them as participants to an assignment when they don't already have an account, e-mail is not sent. Students should receive e-mails upon account creation, regardless of how their account is created. So this involves adding a call to the e-mailer … or, perhaps, moving an email call from the Users controller to the point where an account is actually created.

Modified requirement : After discussion with the Professor, the requirement was modified to the below :
If someone attempts to assign a nonexistent user as a participant in an assignment by filling out the form on the Add Participants page, (s)he should be warned that the user does not exist.  This is reasonable behaviour because the username may have been mistyped, and you wouldn't want to create a new user account due to a typo.

How to test : 
Adding a non-existent user as a participant in an assignment
1. Login as an instructor.
2. Click on an existing active Assignment .
3. Click on Add Participants and add any non-existent user ( any email id ).
Result : It gives a flash message regarding user does not exist and gives you the link to create a new user.

******************************************************************************************

Requirement 2 : 
Requirement : E-mails about reviews should direct the user to the page where the review is found.

How to test:
1. Login as an instructor.
2. Click on an existing active assignment.
3. Add an existing user as a reviewer and another user as a participant to this assignment.
4. Login as the reviewer and submit your review. 
Result: The participant must have received an email on this review submission with a link to the review submitted.

******************************************************************************************

Requirement 3 :
Original Requirement : Deadline reminders should include a link on where to go to perform the needed function.

Modified Requirement : As there are no deadline reminders implement in expertiza currently , we were asked to verify  the Rspec test cases which are already present in delayed_mailer_spec.rb for the same . We were asked to fix if any bugs were found in that. But no bugs were found.

How to test :
1. Run the Rspec test cases by  following a command :
bundle exec rspec delayed_mailer_spec.rb
Result : See the attachment for the screenshot.

******************************************************************************************

Requirement 4:
Original Requirement : Instructor notifications of where reviews disagree by more than a threshold # of points should point the instructor to the reviews that disagree.
Modified Requirement :The original requirement was already implemented, therefore the Professor suggested the below possible improvements :-
•	Put the links in the bulleted list that is easy to read.
•	The mail should have the previous average score as well as the new score that is being assigned.
•	Any other kind of improvement to increase the readability of the email is welcome.

How to test :
1. Login as an instructor and create three students. 
2. Hover over Manage Tab --> click on Assignments --> choose an existing active assignment-->Click on Add Participants section of that assignment.
3. Out of the three users that you have created, add two as reviewers and one as participant to that assignment.
4. Login as the participant and submit any link as your work .
5. After the submission deadline is over, login as each of the reviewers and submit two contradicting reviews of the same assignment. ( Say, one gives all 0s , another gives all 5s).
Result : This should send an email to the instructor with the below text.

******************************************************************************************
Requirement 5:
Requirement : Send out an email to the invitee when a participant sends out an invitation to another participant to join a team.
How to test:
1.Login as an instructor --> click on Assignments tab --> Select an existing active assignment .
2.Add one user as a participant , login as the participant and invite a student to join the team for the assignment.
Result: Email has been sent to invitee.

******************************************************************************************
Requirement 6 :
Requirement : The student who issued the invitation should also be e-mailed when the invitee joins the team. And also when a student responds to a teammate advertisement. In general, all activity on ad responses and invitations should be reported to the other party by e-mail (unless these e-mails are turned off in a (new) profile field).
How to test :
1. Continue from Requirement 5. 
Result : Both the inviter and the invitee should get an email as per the requirement specified above.

******************************************************************************************

Requirement 7:
Original Requirement : Notify an instructor by e-mail when a student suggests a topic.
How to test :
1. Login as an instructor--> Create a user with Student role
2. Login as a student--> Click on existing active Assignment .
3. Click on suggest a topic handle and suggest any topic and save it
Result : Instructor is notified when the topic is suggested.

*****************************************************************************************

Requirement 8:
Original Requirement : Create an option (in the instructor’s profile) to get a copy of ‘e-mails being sent to students (this is so the instructor can verify correct functioning of the system).
Modified Requirement : Implemented for one such scenario, can easily be replicated for others.

How to test :
1. Login as an existing instructor --> go to profile tab.
2  Do not tick the check box mentioning "get copy of all emails"  now. 
3. Go to Manage Users --> Create a student.
 
Result 1 (negative testing) : Instructor does not receive an email.

1. Login as the instructor and go to the Profile tab.
2. Tick the check box mentioning "get copy of all emails"  now and Save.
3. Perform an event. Here, "Create any other user (may be a Student , Reader ) and Save. While creating the user , please ensure you give an email id that you can access.

Result2(positive testing) : Both the created user and the instructor receives the email.
Note : The instructor only receives copy of all emails happening in the system, if the check box is checked. In this case, we have just implemented for one such scenario . This can easily be replicated for the others.
******************************************************************************************
